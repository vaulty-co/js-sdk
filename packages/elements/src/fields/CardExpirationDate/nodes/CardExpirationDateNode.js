import IMask from 'imask';
import { TextNode } from '../../../nodes/TextNode';

/**
 * Create card expiration date node
 * @class
 */
class CardExpirationDateNode extends TextNode {
  /**
   * @param {TextNodeOptions} [options = {}]
   */
  constructor(options = {}) {
    super(options);

    this.patternMask = IMask(this.node, {
      mask: '00/00',
      lazy: false,
    });
  }

  /**
   * Get value
   */
  getValue() {
    return this.patternMask.unmaskedValue;
  }

  destroy() {
    super.destroy();

    if (this.patternMask) {
      this.patternMask.destroy();
      this.patternMask = null;
    }
  }
}

export default CardExpirationDateNode;
export {
  CardExpirationDateNode,
};
