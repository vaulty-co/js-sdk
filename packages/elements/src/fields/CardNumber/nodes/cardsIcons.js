import visa from './images/visa.svg';
import mastercard from './images/mastercard.svg';
import americanExpress from './images/amex.svg';
import dinersClub from './images/diners.svg';
import discover from './images/discover.svg';
import jcb from './images/jcb.svg';
import unionpay from './images/unionpay.svg';
import maestro from './images/maestro.svg';
import mir from './images/mir.svg';
import elo from './images/elo.svg';
import hiper from './images/hiper.svg';
import hipercard from './images/hipercard.svg';
import generic from './images/generic.svg';

const cardsIcons = {
  visa,
  mastercard,
  'american-express': americanExpress,
  'diners-club': dinersClub,
  discover,
  jcb,
  unionpay,
  maestro,
  mir,
  elo,
  hiper,
  hipercard,
  generic,
};

export default cardsIcons;
export {
  cardsIcons,
};
