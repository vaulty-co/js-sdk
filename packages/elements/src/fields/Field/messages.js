const IS_MOUNTED_REQUEST = 'is_mounted_request';
const IS_MOUNTED_RESPONSE = 'is_mounted_response';

export {
  IS_MOUNTED_REQUEST,
  IS_MOUNTED_RESPONSE,
};
