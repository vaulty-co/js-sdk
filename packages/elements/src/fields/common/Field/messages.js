const INITIALIZE_FIELD_REQUEST = 'initialize_request';
const INITIALIZE_FIELD_RESPONSE = 'initialize_response';
const GET_FIELD_DATA_REQUEST = 'get_field_data_request';
const GET_FIELD_DATA_RESPONSE = 'get_field_data_response';
const FIELD_DATA_CHANGE_RESPONSE = 'FIELD_DATA_CHANGE_RESPONSE';

export {
  INITIALIZE_FIELD_REQUEST,
  INITIALIZE_FIELD_RESPONSE,
  GET_FIELD_DATA_REQUEST,
  GET_FIELD_DATA_RESPONSE,
  FIELD_DATA_CHANGE_RESPONSE,
};
