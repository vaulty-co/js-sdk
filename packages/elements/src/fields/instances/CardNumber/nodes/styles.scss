.cardNumberWrapper {
  display: flex;
  width: 100%;
  height: 100%;
}

.cardNumberIcon {
  display: block;
  height: 100%;
  margin-right: 0.3em;
}

.cardNumberInput {
  display: block;
  width: 100%;
  height: 100%;
  font-size: 100%;
  flex-grow: 1;
  border: 0 none;
  outline: 0 none;
  box-sizing: border-box;
  background: transparent;
}

.cardNumberInput:disabled {
  cursor: not-allowed;
}
