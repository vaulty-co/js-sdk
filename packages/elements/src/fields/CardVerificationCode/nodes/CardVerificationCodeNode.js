import IMask from 'imask';
import { TextNode } from '../../../nodes/TextNode';

/**
 * Create card verification code node
 * @class
 */
class CardVerificationCodeNode extends TextNode {
  /**
   * @param {Object} [options = {}]
   * @param {string} [options.type = 'text']
   */
  constructor(options = {}) {
    super(options);

    this.patternMask = IMask(this.node, {
      mask: '000[0]',
    });
  }

  /**
   * Get value
   */
  getValue() {
    return this.patternMask.unmaskedValue;
  }

  destroy() {
    super.destroy();

    if (this.patternMask) {
      this.patternMask.destroy();
      this.patternMask = null;
    }
  }
}

export default CardVerificationCodeNode;
export {
  CardVerificationCodeNode,
};
