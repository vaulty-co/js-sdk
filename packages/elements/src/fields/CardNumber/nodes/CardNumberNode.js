import IMask from 'imask';
import cardValidator from 'card-validator';
import { Node } from '@js-sdk/utils/src/nodes/Node';

import styles from './styles.scss';
import { cardsIcons } from './cardsIcons';

const timesString = (count, character) => (
  Array
    .from({ length: count })
    .reduce((resultString) => `${resultString}${character}`, '')
);

const getMask = (card) => {
  const cardLengthMin = Math.min(...card.lengths);
  const cardLengthMax = Math.max(...card.lengths);

  let maskWithoutGaps = timesString(cardLengthMin, '0');
  maskWithoutGaps += `[${timesString(cardLengthMax - cardLengthMin, '0')}]`;
  maskWithoutGaps = maskWithoutGaps.replace('[]', '');

  return [
    ...card.gaps,
    maskWithoutGaps.length,
  ]
    .reduce(
      (data, gap) => ({
        previousCursor: gap,
        result: [
          ...data.result,
          maskWithoutGaps.substring(data.previousCursor, gap),
        ],
      }),
      {
        previousCursor: 0,
        result: [],
      },
    )
    .result
    .join(' ');
};

const STYLES_FOR_WRAPPER = ['width', 'height'];
const getStyleForInput = (style = {}) => (
  Object
    .keys(style)
    .reduce((resultStyles, name) => {
      if (!STYLES_FOR_WRAPPER.includes(name)) {
        return {
          ...resultStyles,
          [name]: style[name],
        };
      }
      return resultStyles;
    }, {})
);

/**
 * Create card number node with using <input /> element
 * @class
 */
class CardNumberNode extends Node {
  /**
   * @param {Object} [options = {}]
   */
  constructor(options = {}) {
    const wrapper = document.createElement('div');
    wrapper.className = styles.cardNumberWrapper;
    wrapper.innerHTML = `
        <img class="${styles.cardNumberIcon}" src="${cardsIcons.generic}" alt="" />
        <input class="${styles.cardNumberInput}" type="tel" />
    `;

    super({
      ...options,
      node: wrapper,
    });

    this.input = wrapper.querySelector('input');
    this.image = wrapper.querySelector('img');

    this.patternMask = IMask(this.input, {
      mask: '0000 0000 0000 0000',
      lazy: false,
    });

    this.handleInput = this.handleInput.bind(this);
    this.on('input', this.handleInput);
  }

  /**
   * Add event listener
   * @param {string} eventName
   * @param {function(e:Event)} handler
   */
  on(eventName, handler) {
    if (['focus', 'blur'].includes(eventName)) {
      this.input.addEventListener(eventName, handler);
    } else {
      super.on(eventName, handler);
    }
  }

  /**
   * Remove event listener
   * @param {string} eventName
   * @param {function(e:Event)} handler
   */
  off(eventName, handler) {
    if (['focus', 'blur'].includes(eventName)) {
      this.input.removeEventListener(eventName, handler);
    } else {
      super.off(eventName, handler);
    }
  }

  /**
   * Get text input value
   */
  getValue() {
    return this.patternMask.unmaskedValue;
  }

  /**
   * Handle input change
   */
  handleInput() {
    const value = this.getValue();

    const cardData = cardValidator.number(value);
    const card = cardData?.card ?? { lengths: [16], gaps: [4, 8, 12], type: 'generic' };

    this.patternMask.updateOptions({
      mask: getMask(card),
    });

    this.image.src = cardsIcons[card.type];
  }

  /**
   * Set style to node
   * @param {Object} [style = {}]
   */
  setStyle(style) {
    const styleForInput = getStyleForInput(style);
    Object.assign(
      this.node,
      style,
    );
    Object.assign(
      this.input,
      styleForInput,
    );
  }

  /**
   * Destroy component
   */
  destroy() {
    this.off('input', this.handleInput);

    this.patternMask.destroy();
    this.patternMask = null;

    super.destroy();

    this.input = null;
    this.image = null;
  }
}

export default CardNumberNode;
export {
  CardNumberNode,
  getMask,
};
