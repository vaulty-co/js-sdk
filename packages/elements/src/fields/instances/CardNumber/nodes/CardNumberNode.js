import IMask from 'imask';
import cardValidator from 'card-validator';
import { Node } from '@vaulty/common/src/nodes/Node';

import styles from './styles.scss';
import { cardsIcons } from './cardsIcons';

const timesString = (count, character) => (
  Array
    .from({ length: count })
    .reduce((resultString) => `${resultString}${character}`, '')
);

const getMask = (card) => {
  const cardLengthMin = Math.min(...card.lengths);
  const cardLengthMax = Math.max(...card.lengths);

  let maskWithoutGaps = timesString(cardLengthMin, '0');
  maskWithoutGaps += `[${timesString(cardLengthMax - cardLengthMin, '0')}]`;
  maskWithoutGaps = maskWithoutGaps.replace('[]', '');

  return [
    ...card.gaps,
    maskWithoutGaps.length,
  ]
    .reduce(
      (data, gap) => ({
        previousCursor: gap,
        result: [
          ...data.result,
          maskWithoutGaps.substring(data.previousCursor, gap),
        ],
      }),
      {
        previousCursor: 0,
        result: [],
      },
    )
    .result
    .join(' ');
};

const STYLES_FOR_WRAPPER = ['width', 'height'];
const getStyleForInput = (style = {}) => (
  Object
    .keys(style)
    .reduce((resultStyles, name) => {
      if (!STYLES_FOR_WRAPPER.includes(name)) {
        return {
          ...resultStyles,
          [name]: style[name],
        };
      }
      return resultStyles;
    }, {})
);

/**
 * Create card number node with using <input /> element
 * @class
 */
class CardNumberNode extends Node {
  /**
   * @param {Object} [options = {}]
   */
  constructor(options = {}) {
    const wrapper = document.createElement('div');
    wrapper.className = styles.cardNumberWrapper;
    wrapper.innerHTML = `
        <img class="${styles.cardNumberIcon}" src="${cardsIcons.generic}" alt="" />
        <input class="${styles.cardNumberInput}" type="tel" />
    `;

    super({
      ...options,
      node: wrapper,
    });

    this.input = wrapper.querySelector('input');
    this.image = wrapper.querySelector('img');

    this.patternMask = IMask(this.input, {
      mask: '0000 0000 0000 0000',
      lazy: false,
    });

    this.handleInput = this.handleInput.bind(this);
    this.on('input', this.handleInput);
  }

  /**
   * Add event listener
   * @param {string} eventName
   * @param {function(e:Event)} handler
   */
  on(eventName, handler) {
    if (['focus', 'blur'].includes(eventName)) {
      this.input.addEventListener(eventName, handler);
    } else {
      super.on(eventName, handler);
    }
  }

  /**
   * Focus card number node
   */
  focus() {
    this.input.focus();
  }

  /**
   * Blur card number node
   */
  blur() {
    this.input.blur();
  }

  /**
   * Clear card number node
   */
  clear() {
    this.patternMask.value = '';
    this.patternMask.updateControl();
  }

  /**
   * Remove event listener
   * @param {string} eventName
   * @param {function(e:Event)} handler
   */
  off(eventName, handler) {
    if (['focus', 'blur'].includes(eventName)) {
      this.input.removeEventListener(eventName, handler);
    } else {
      super.off(eventName, handler);
    }
  }

  /**
   * Get text input value
   */
  getValue() {
    return this.patternMask.unmaskedValue;
  }

  /**
   * Set style to node
   * @param {Object} [style = {}]
   */
  setStyle(style) {
    const styleForInput = getStyleForInput(style);
    Object.assign(
      this.node,
      style,
    );
    Object.assign(
      this.input,
      styleForInput,
    );
  }

  /**
   * Set attributes to input
   * @param {FieldAttributes} fieldAttributes
   */
  setAttributes(fieldAttributes) {
    if (fieldAttributes.disabled) {
      this.input.setAttribute('disabled', '');
    } else {
      this.input.removeAttribute('disabled');
    }
  }

  /**
   * Destroy component
   */
  destroy() {
    this.off('input', this.handleInput);

    this.patternMask.destroy();
    this.patternMask = null;

    super.destroy();

    this.input = null;
    this.image = null;
  }

  /**
   * Handle input change
   * @private
   */
  handleInput() {
    const value = this.getValue();

    const cardData = cardValidator.number(value);
    const card = cardData?.card ?? { lengths: [16], gaps: [4, 8, 12], type: 'generic' };

    this.patternMask.updateOptions({
      mask: getMask(card),
    });

    this.image.src = cardsIcons[card.type];
  }
}

export default CardNumberNode;
export {
  CardNumberNode,
  getMask,
};
