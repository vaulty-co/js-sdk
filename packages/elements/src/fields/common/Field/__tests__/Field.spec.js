import { enforceOptions } from 'broadcast-channel';
import { Node } from '@vaulty/common/src/nodes/Node';
import { Message } from '@vaulty/common/src/channels/Message';

import { Field } from '../index';
import { FIELD_LOADED_WATCHER } from '../messages';

jest.mock('@vaulty/common/src/channels/SlaveChannel', () => {
  const connectMock = jest.fn();
  const subscribeMock = jest.fn();
  const postMessageMock = jest.fn();
  const destroyMock = jest.fn();

  class SlaveChannel {
    get connect() {
      return connectMock;
    }

    get subscribe() {
      return subscribeMock;
    }

    get postMessage() {
      return postMessageMock;
    }

    destroy() {
      return destroyMock;
    }
  }
  return {
    SlaveChannel,
  };
});

const createNode = () => (
  new Node({
    node: document.createElement('div'),
  })
);

beforeAll(() => {
  enforceOptions({
    type: 'simulate',
  });
});

afterAll(() => {
  enforceOptions(null);
});

describe('Field', () => {
  describe('#constructor', () => {
    it('should throw error, when node is not HTMLELement', () => {
      const fieldNode = {};

      expect(() => {
        new Field({ node: fieldNode, channelId: 'channel-id' });
      }).toThrowErrorMatchingSnapshot();
    });
  });

  describe('#mount', () => {
    let parentNode;
    let field;

    beforeEach(() => {
      field = new Field({
        node: createNode(),
        channelId: 'channel-id',
      });
      parentNode = document.createElement('div');
    });

    afterEach(() => {
      field.destroy();
      field = null;
      parentNode = null;
    });

    it('should throw error, when channelId is not specified', () => {
      field = new Field({ node: createNode() });

      expect(() => {
        field.mount(parentNode);
      }).toThrowErrorMatchingSnapshot();
    });

    it('should append field node to parent node', () => {
      field.mount(parentNode);

      expect(parentNode.firstChild).toBe(field.fieldNode.node);
    });

    it('should initialize slave channel connected to parent channel', () => {
      field.mount(parentNode);

      expect(field.fieldSlaveChannel.connect).toBeCalled();
    });

    it('should post message that field is loaded', () => {
      field.mount(parentNode);

      expect(field.fieldSlaveChannel.postMessage).toBeCalledWith(
        new Message(FIELD_LOADED_WATCHER),
      );
    });

    it('should add handlers to slave channel', () => {
      field.mount(parentNode);

      expect(field.fieldSlaveChannel.subscribe).toBeCalled();
    });
  });

  describe('#appendTo', () => {
    it('should append field node in specified parent node', () => {
      const fieldNode = createNode();
      const parentNode = document.createElement('div');
      const field = new Field({ node: fieldNode, channelId: 'channel-id' });

      field.appendTo(parentNode);

      expect(parentNode.firstChild).toBe(fieldNode.node);
    });

    it('should throw error, when parent node is not HTMLElement', () => {
      const fieldNode = createNode();
      const parentNode = undefined;
      const field = new Field({ node: fieldNode, channelId: 'channel-id' });

      expect(() => {
        field.appendTo(parentNode);
      }).toThrowErrorMatchingSnapshot();
    });
  });

  describe('#destroy', () => {
    let fieldNode;
    let parentNode;
    let field;

    beforeEach(() => {
      fieldNode = createNode();
      parentNode = document.createElement('div');
      field = new Field({ node: fieldNode, channelId: 'channel-id' });

      field.appendTo(parentNode);
    });

    it('should remove node from parent, if field has been appended to its', () => {
      field.destroy();

      expect(parentNode.firstChild).toBe(null);
    });

    it('should clear all references about nodes (parent and field node)', () => {
      field.destroy();

      expect(field.fieldNode).toBe(null);
    });

    it('should throw exception when appendTo is called after destroying', () => {
      field.destroy();

      expect(() => {
        field.appendTo(parentNode);
      }).toThrowErrorMatchingSnapshot();
    });

    it('should throw exception when destroy is called after destroying', () => {
      field.destroy();

      expect(() => {
        field.destroy();
      }).toThrowErrorMatchingSnapshot();
    });
  });
});
