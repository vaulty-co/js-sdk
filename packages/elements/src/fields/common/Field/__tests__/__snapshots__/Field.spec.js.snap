// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field #appendTo should throw error, when parent node is not HTMLElement 1`] = `"[Node] Parent node should be a HTMLElement node"`;

exports[`Field #constructor should throw error, when node is not HTMLELement 1`] = `"[Field] Field's node should be instance of Node"`;

exports[`Field #destroy should throw exception when appendTo is called after destroying 1`] = `"[Field] Field is destroyed and can not be append to a parent node."`;

exports[`Field #destroy should throw exception when destroy is called after destroying 1`] = `"[Field] Field is destroyed and can not be destroyed again."`;

exports[`Field #mount should throw error, when channelId is not specified 1`] = `"[Field] Field does not have specified \\"channelId\\" for opening channel"`;
